sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -1761,16 +1772,21 @@ object CodeGenerator extends Logging {
 
         case ref: BoundReference if ctx.currentVars != null &&
             ctx.currentVars(ref.ordinal) != null =>
-          val ExprCode(_, isNull, value) = ctx.currentVars(ref.ordinal)
-          collectLocalVariable(value)
-          collectLocalVariable(isNull)
+          val exprCode = ctx.currentVars(ref.ordinal)
+          // If the referred variable is not evaluated yet.
+          if (exprCode.code != EmptyBlock) {
+            exprCodesNeedEvaluate += exprCode.copy()
```

Comment:
We need this copy? A unnecessary copy can happen if `exprCodesNeedEvaluate` already has the same entry?

Reply:
Copying `exprCode` because we need the unevaluated `code` in `exprCode`, but we also need to empty `code` of `exprCode`. We need copied code so we can evaluate them before evaluating subexpressions. We need to empty `code` of `exprCode` so we don't re-evaluate the code.

Reply:
> We need to empty code of exprCode so we don't re-evaluate the code.

AFAIK when we empty the code, we also do a copy, right?

Reply:
`exprCode.code = EmptyBlock`, do you mean this?

Reply:
Sorry I was wrong, the copy here is necessary.